doc/opannotate.1.in
.TH OPANNOTATE 1 "@DATE@" "oprofile @VERSION@"
.UC 4
.SH NAME
opannotate \- produce source or assembly annotated with profile data
.SH SYNOPSIS
.br
.B opannotate
[
.I options
]
[profile specification]
.SH DESCRIPTION

.B opannotate
outputs annotated source and/or assembly from profile data of an OProfile
session.

.SH OPTIONS
.TP
.BI "--help / -? / --usage"
Show help message.
.br
.TP
.BI "--version / -v"
Show version.
.br
.TP
.BI "--verbose / -V"
Give verbose debugging output.
.br
.TP
.BI "--image-path / -p [paths]"
Comma-separated list of additional paths to search for binaries.
This is needed to find modules in kernels 2.5 and upwards.
.br
.TP
.BI "--threshold / -t [percentage]"
Only output data for symbols that have more than the given percentage
of total samples.
.br
.TP
.BI "--no-demangle"
Do not attempt to demangle C++ symbols.
.br
.TP
.BI "--smart-demangle / -D"
Use pattern-matching to make C++ symbol demangling more readable.
.br
.TP
.BI "--source-dir / -d [dir]"
Top source directory for the source files.
.br
.TP
.BI "--output-dir / -o [dir]"
Output directory. This makes opannotate output one annotated file for each
source file.
.br
.TP
.BI "--base-dir / -b [dir]"
FIXME
.br
.TP
.BI "--include-file [files]"
Only include files in the given comma-separated list of glob patterns.
.br
.TP
.BI "--exclude-file [files]"
Exclude all files in the given comma-separated list of glob patterns.
.br
.TP
.BI "--include-symbols [symbols]"
Only include symbols in the given comma-separated list.
.br
.TP
.BI "--exclude-symbols [symbols]"
Exclude all the symbols in the given comma-separated list.
.br
.TP
.BI "--objdump-params [params]"
Pass the given parameters as extra values when calling objdump.
.br
.TP
.BI "--exclude-dependent"
Do not include application-specific images for libraries, kernel modules
and the kernel. This option only makes sense if the profile session
used --separate.
.br
.TP
.BI "--source / -s"
Output annotated source. This requires debugging information to be available
for the binaries.
.br
.TP
.BI "--assembly / -a"
Output annotated assembly. If this is combined with --source, then mixed
source / assembly annotations are output.

.SH ENVIRONMENT
No special environment variables are recognised by opannotate.

.SH FILES
.TP
.I /var/lib/oprofile/samples/
The location of the generated sample files.

.SH VERSION
.TP
This man page is current for @PACKAGE@-@VERSION@.

.SH SEE ALSO
.BR oprofile(1)


ChangeLog
2003-07-05  Philippe Elie  <phil.el@wanadoo.fr>

	* daemon/opd_image.c:
	* libdb/db_insert.c: 64 bits arch compile fix, thanks to Falk Hueffner
	 <falk@debian.org>

2003-07-02  Philippe Elie  <phil.el@wanadoo.fr>

	* pp/opreport.cpp: honor options::show_header

2003-06-30  Graydon Hoare  <graydon@redhat.com>

	* libpp/op_header.cpp: Handle printing multiple bitmask values.

2003-06-19  John Levon  <levon@movementarian.org>

	* doc/Makefile.am: fix make distcheck

2003-06-18  John Levon  <levon@movementarian.org>

	* pp/opannotate_options.c: hide the unhandled
	--base-dir option

2003-06-18  John Levon  <levon@movementarian.org>

	* pp/Makefile.am: don't build opdiff

2003-06-17  John Levon  <levon@movementarian.org>

	* utils/opcontrol: don't accept "-r" - it's too easy to typo

2003-06-17  Will Cohen  <wcohen@redhat.com>

	* daemon/opd_cookie.h: Add s390 syscall value.
	
2003-06-17  John Levon  <levon@movementarian.org>

	* utils/opcontrol: don't accept "-r" - it's too easy to typo

2003-06-17  John Levon  <levon@movementarian.org>

	* doc/srcdoc/Doxyfile: removed

	* doc/srcdoc/Doxyfile.in:
	* configure.in: generate doxygen version nr. automatically

2003-06-17  Philippe Elie  <phil.el@wanadoo.fr>

	* libdb/db_insert.cpp: missing initialization of error message on
	  error path

2003-06-17  John Levon  <levon@movementarian.org>

	* utils/opcontrol: accept some short forms for common
	operations

2003-06-17  John Levon  <levon@movementarian.org>

	* utils/opcontrol: accept --foo blah as well as
	--foo=blah

2003-06-16  John Levon  <levon@movementarian.org>

	* utils/opcontrol: "opcontrol" will show usage
	instead of doing nothing

2003-06-16  Will Cohen  <wcohen@nc.rr.com>

	* libabi/abi.cpp:
	* libabi/op_import.cpp: Add required include for cassert.

2003-06-16  John Levon  <levon@movementarian.org>

	* events/x86-64.hammer.unit_masks: remove some
	bogus FIXMEs

2003-06-15  John Levon  <levon@movementarian.org>

	* README: fix autogen.sh line

2003-06-15  John Levon  <levon@movementarian.org>

	* daemon/opd_image.c: 64 bit fixes

2003-06-15  John Levon  <levon@movementarian.org>

	* utils/op_help.c: c89 fix

2003-06-15  John Levon  <levon@movementarian.org>

	* daemon/opd_image.c: c89 fix

2003-06-15  John Levon  <levon@movementarian.org>

	* doc/oprofile.xml: document --event

	* utils/opcontrol:
	* utils/op_help.c: allow unit mask, kernel, user to be
	optional. Verify not too many events are passed.

2003-06-15  John Levon  <levon@movementarian.org>

	* libop/op_cpu_type.c: fix nr_counters cpu type check

	* utils/op_help: implement --check_events, event validation
	and counter allocation

	* utils/opcontrol: use the above, implement --event=default

2003-06-14  John Levon  <levon@movementarian.org>

	* daemon/opd_image.c: fix a stupid error in my previous TGID patch

2003-06-12  Graydon Hoare  <graydon@redhat.com>

	* libop/op_events.c (op_check_events): Check bitmasks as well as 
	exclusive values, for unit masks.

2003-06-12  John Levon  <levon@movementarian.org>

	* daemon/opd_image.c: we were not handling a truncated
	TGID entry properly

2003-06-12  John Levon  <levon@movementarian.org>

	* libutil++/unique_storage.h: missing typename

2003-06-11  Philippe Elie  <phil.el@wanadoo.fr>

	* pp/opannotate_options.cpp: typo in option name

2003-06-11  Philippe Elie  <phil.el@wanadoo.fr>

	* libutil++/op_bfd.cpp: boring_symbol() new to get a better choice
	when eliminatiing symbol at identical vma

2003-06-11  Philippe Elie  <phil.el@wanadoo.fr>

	* utils/opcontrol: fix rtc option checking, bug added after 0.5.3,
	no big deal

2003-06-11  Philippe Elie  <phil.el@wanadoo.fr>

	* libregex/op_regex.cpp:
	* libpp/profile.cpp: compile fix, tree was broken by my previous patch

2003-06-11  Jason Yeh  <jason.yeh@amd.com>

	* events/i386.athlon.events: add CPU_CLK_UNHALTED

	* utils/opcontrol: use the above by default

	* events/x86-64.hammer.events:
	* events/x86-64.hammer.unit_masks: various fixes

2003-06-11  Philippe Elie  <phil.el@wanadoo.fr>

	* libpp/op_header.cpp:
	* libpp/profile.cpp:
	* libpp/profile_spec.cpp:
	* libpp/symbol_sort.cpp:
	* libregex/op_regex.cpp:
	* libutil++/op_bfd.cpp: use exception rather exit(EXIT_FAILURE) in
	library code

2003-06-11  John Levon  <levon@movementarian.org>

	* daemon/opd_image.c: future-proof the code handlers

2003-06-11  Philippe Elie  <phil.el@wanadoo.fr>

	* pp/opannotate.cpp: output_asm() avoid output when the set of selected
	symbols doesn't contain any samples

2003-06-11  Philippe Elie  <phil.el@wanadoo.fr>

	* libutil++/unique_storage.h: we don't need to check if key is present
	  before trying to insert it

2003-06-10  Philippe Elie  <phil.el@wanadoo.fr>

	* libpp/name_storage.h: minor tidy
	* libutil++/unique_storage.h: small efficiency improvement

2003-06-09  John Levon  <levon@movementarian.org>

	* pp/Makefile.am:
	* libutil/Makefile.am: fix make distcheck

2003-06-09  John Levon  <levon@movementarian.org>

	* libpp/opp_symbol.h: rename to ...

	* libpp/symbol.h: ... this

	* libpp/Makefile.am:
	* libpp/format_output.h:
	* libpp/format_output.cpp:
	* libpp/profile_container.h:
	* libpp/profile_container.cpp:
	* libpp/sample_container.h:
	* libpp/sample_container.cpp:
	* libpp/symbol_functors.h:
	* libpp/symbol_sort.cpp: fix up from above and some header
	cleanups

2003-06-09  John Levon  <levon@movementarian.org>

	* libpp/profile.h: std::pair not pair

2003-06-08  Philippe Elie  <phil.el@wanadoo.fr>

	* libpp/profile.cpp: don't underflow start offset when the vma range
	  is unknown
	* pp/opreport.cpp: use samples_range() not samples_range(0, ~0)

2003-06-08  Philippe Elie  <phil.el@wanadoo.fr>

	* libpp/profile.h:
	* libpp/profile.cpp: replace accumulate_samples() API by an iterator
	  interface
	* libpp/profile_container.h:
	* libpp/profilecontainer.cpp:
	* pp/opreport.cpp: use new API

2003-06-08  Philippe Elie  <phil.el@wanadoo.fr>

	* include/sstream: replace old version by the once shipped with

pp/opannotate.cpp
/**
 * @file opannotate.cpp
 * Implement opannotate utility
 *
 * @remark Copyright 2003 OProfile authors
 * @remark Read the file COPYING
 *
 * @author John Levon
 * @author Philippe Elie
 */

#include <iostream>
#include <sstream>
#include <algorithm>
#include <iomanip>
#include <fstream>
#include <utility>

#include "op_header.h"
#include "profile.h"
#include "op_sample_file.h"
#include "cverb.h"
#include "string_manip.h"
#include "demangle_symbol.h"
#include "child_reader.h"
#include "op_file.h"
#include "file_manip.h"
#include "partition_files.h"
#include "opannotate_options.h"
#include "profile_container.h"

using namespace std;
using namespace options;

namespace {

scoped_ptr<profile_container> samples;

// needed to display samples file information
scoped_ptr<opd_header> header;

/// how opannotate was invoked
string cmdline;

/// empty annotation fill string
string annotation_fill;

/// string used as start / end comment to annotate source
string const begin_comment("/* ");
string const in_comment(" * ");
string const end_comment(" */");

/// field width for the sample count
unsigned int const count_width = 6;


// FIXME share with opgprof.cpp and opreport.cpp
image_set populate_samples(profile_container & samples,
			   partition_files const & files,
			   bool merge_lib)
{
	image_set images = sort_by_image(files, extra_found_images);

	image_set::const_iterator it;
	for (it = images.begin(); it != images.end(); ) {
		pair<image_set::const_iterator, image_set::const_iterator>
			p_it = images.equal_range(it->first);

		if (p_it.first == p_it.second)
			continue;

		op_bfd abfd(p_it.first->first, symbol_filter);
		profile_t profile;

		string app_name = p_it.first->second.image;
		if (merge_lib) {
			app_name = p_it.first->first;
		}

		for (it = p_it.first;  it != p_it.second; ++it) {
			profile.add_sample_file(it->second.sample_filename,
						abfd.get_start_offset());
		}

		check_mtime(abfd.get_filename(), profile.get_header());
	
		samples.add(profile, abfd, app_name);
	}

	return images;
}

void save_sample_file_header(partition_files const & files)
{
	if (files.nr_set()) {
		partition_files::filename_set const & file_set = files.set(0);
		opd_header temp = read_header(file_set.begin()->sample_filename);
		header.reset(new opd_header(temp));
	}
}


string get_annotation_fill()
{
	string str;

	str += string(count_width, ' ') + ' ';
	str += string(percent_width, ' '); 

	return str;
}


symbol_entry const * find_symbol(string const & image_name,
				 string const & str_vma)
{
	// do not use the bfd equivalent:
	//  - it does not skip space at begin
	//  - we does not need cross architecture compile so the native
	// strtoull must work, assuming unsigned long long can contain a vma
	// and on 32/64 bits box bfd_vma is 64 bits
	bfd_vma vma = strtoull(str_vma.c_str(), NULL, 16);

	return samples->find_symbol(image_name, vma);
}


void output_info(ostream & out)
{
	out << begin_comment << endl;

	out << in_comment << "Command line: " << cmdline << endl
	    << in_comment << endl;

	out << in_comment << "Interpretation of command line:" << endl;

	if (!assembly) {
		out << in_comment
		    << "Output annotated source file with samples" << endl;

		// FIXME: re-add this
#if 0
		if (threshold_percent != 0) {
			if (!do_until_more_than_samples) {
				out << in_comment
				    << "Output files where the selected counter reach "
				    << threshold_percent << "% of the samples"
				    << endl;
			} else {
				out << in_comment << "output files until "
				    << threshold_percent
				    << "% of the samples is reached on the selected counter"
				    << endl;
			}
		} else
#endif
		{
			out << in_comment << "Output all files" << endl;
		}
	} else {
		out << in_comment
		    << "Output annotated assembly listing with samples"
		    << endl;

		if (!objdump_params.empty()) {
			out << in_comment << "Passing the following "
				"additional arguments to objdump ; \"";
			for (size_t i = 0 ; i < objdump_params.size() ; ++i)
				out << objdump_params[i] << " ";
			out << "\"" << endl;
		}
	}

	out << in_comment << endl;

	stringstream stream;

	stream << *header;
	stream.seekp(0);

	string line;
	while (getline(stream, line)) {
		out << in_comment << line << endl;
	}

	out << end_comment << endl;
}


string counter_str(size_t counter, size_t total)
{
	ostringstream os;
	os << setw(count_width) << counter << ' ';

	os << format_double(op_ratio(counter, total) * 100.0,
			    percent_int_width, percent_fract_width);
	return os.str();
}


string asm_line_annotation(symbol_entry const * last_symbol,
			   string const & value)
{
	// do not use the bfd equivalent:
	//  - it does not skip space at begin
	//  - we does not need cross architecture compile so the native
	// strtoull must work, assuming unsigned long long can contain a vma
	// and on 32/64 bits box bfd_vma is 64 bits
	bfd_vma vma = strtoull(value.c_str(), NULL, 16);

	string str;

	sample_entry const * sample = samples->find_sample(last_symbol, vma);
	if (sample)
		str += counter_str(sample->count, samples->samples_count());

	if (str.empty())
		str = annotation_fill;

	str += " :";
	return str;
}


string symbol_annotation(symbol_entry const * symbol)
{
	if (!symbol)
		return string();

	string annot = counter_str(symbol->sample.count,
				   samples->samples_count());
	if (annot.empty())
		return  string();

	string symname = demangle_symbol(symbol->name);

	string str = " ";
	str += begin_comment + symname + " total: ";
	str += counter_str(symbol->sample.count, samples->samples_count());
	str += end_comment;
	return str;
}


namespace {

/// return true if  this line contains a symbol name in objdump formatting
/// symbol are on the form 08030434 <symbol_name>:  we need to be strict
/// here to avoid any interpretation of a source line as a symbol line
bool is_symbol_line(string const & str, string::size_type pos)
{
	if (str[pos] != ' ' || str[pos + 1] != '<')
		return false;

	return str[str.length() - 1] == ':';
}

}


symbol_entry const * output_objdump_asm_line(symbol_entry const * last_symbol,
		string const & app_name, string const & str,
		vector<symbol_entry const *> const & output_symbols,
		bool & do_output)
{
	// output of objdump is a human readable form and can contain some
	// ambiguity so this code is dirty. It is also optimized a little bit
	// so it is difficult to simplify it without breaking something ...

	// line of interest are: "[:space:]*[:xdigit:]?[ :]", the last char of
	// this regexp dis-ambiguate between a symbol line and an asm line. If
	// source contain line of this form an ambiguity occur and we rely on
	// the robustness of this code.

	size_t pos = 0;
	while (pos < str.length() && isspace(str[pos]))
		++pos;

	if (pos == str.length() || !isxdigit(str[pos])) {
		if (do_output) {
			cout << annotation_fill << " :" << str << endl;
			return last_symbol;
		}
	}

	while (pos < str.length() && isxdigit(str[pos]))
		++pos;

	if (pos == str.length() || (!isspace(str[pos]) && str[pos] != ':')) {
		if (do_output) {
			cout << annotation_fill << " :" << str << endl;
			return last_symbol;
		}
	}

	if (is_symbol_line(str, pos)) {
		last_symbol = find_symbol(app_name, str);

		// ! complexity: linear in number of symbol must use sorted
		// by address vector and lower_bound ?
		// Note this use a pointer comparison. It work because symbols
		// pointer are unique
		if (find(output_symbols.begin(),
		       output_symbols.end(), last_symbol) != output_symbols.end()) {
			do_output = true;
		} else {
			do_output = false;
		}

		if (do_output)
			cout << str << symbol_annotation(last_symbol) << endl;

	} else {
		// not a symbol, probably an asm line.
		if (do_output)
			cout << asm_line_annotation(last_symbol, str)
			     << str << endl;
	}

	return last_symbol;
}


void do_one_output_objdump(vector<symbol_entry const *> const & output_symbols,
			   string const & app_name, bfd_vma start, bfd_vma end)
{
	vector<string> args;

	args.push_back("-d");
	args.push_back("--no-show-raw-insn");
	if (source)
		args.push_back("-S");

	if (start || end != ~(bfd_vma)0) {
		ostringstream arg1, arg2;
		arg1 << "--start-address=" << start;
		arg2 << "--stop-address=" << end;
		args.push_back(arg1.str());
		args.push_back(arg2.str());
	}

	if (!objdump_params.empty()) {
		for (size_t i = 0 ; i < objdump_params.size() ; ++i)
			args.push_back(objdump_params[i]);
	}

	args.push_back(app_name);
	child_reader reader("objdump", args);
	if (reader.error()) {
		cerr << "An error occur during the execution of objdump:\n\n";
		cerr << reader.error_str() << endl;
		return;
	}

	// to filter output of symbols (filter based on command line options)
	bool do_output = true;

	symbol_entry const * last_symbol = 0;
	string str;
	while (reader.getline(str)) {
		last_symbol = output_objdump_asm_line(last_symbol, app_name,
					str, output_symbols, do_output);
	}

	// objdump always returns SUCCESS so we must rely on the stderr state
	// of objdump. If objdump error message is cryptic our own error
	// message will be probably also cryptic
	ostringstream std_err;
	ostringstream std_out;
	reader.get_data(std_out, std_err);
	if (std_err.str().length()) {
		cerr << "An error occur during the execution of objdump:\n\n";
		cerr << std_err.str() << endl;
		return ;
	}

	reader.terminate_process();  // force error code to be acquired

	// required because if objdump stop by signal all above things suceeed
	// (signal error message are not output through stdout/stderr)
	if (reader.error()) {
		cerr << "An error occur during the execution of objdump:\n\n";
		cerr << reader.error_str() << endl;
		return;
	}
}


void output_objdump_asm(vector<symbol_entry const *> const & output_symbols,
			string const & app_name)
{
	// this is only an optimisation, we can either filter output by
	// directly calling objdump and rely on the symbol filtering or
	// we can call objdump with the right parameter to just disassemble
	// the needed part. This is a real win only when calling objdump
	// a medium number of times, I dunno if the used threshold is optimal
	// but it is a conservative value.
	size_t const max_objdump_exec = 50;
	if (output_symbols.size() <= max_objdump_exec) {
		for (size_t i = 0 ; i < output_symbols.size() ; ++i) {
			bfd_vma start = output_symbols[i]->sample.vma;
			bfd_vma end  = start + output_symbols[i]->size;
			do_one_output_objdump(output_symbols, app_name,
					      start, end);
		}
	} else {
		do_one_output_objdump(output_symbols, app_name, 0, ~(bfd_vma)0);
	}
}


void output_asm(string const & app_name)
{
	profile_container::symbol_choice choice;
	choice.threshold = options::threshold;
	choice.image_name = app_name;
	choice.match_image = true;
	vector<symbol_entry const *> symbols = samples->select_symbols(choice);

	output_info(cout);

	output_objdump_asm(symbols, app_name);
}


string const source_line_annotation(string const & filename, size_t linenr)
{
	string str;

	u32 count = samples->samples_count(filename, linenr);
	if (count)
		str += counter_str(count, samples->samples_count());

	if (str.empty())
		str = annotation_fill;

	str += " :";
	return str;
}


string source_symbol_annotation(string const & filename, size_t linenr)
{
	symbol_entry const * symbol = samples->find_symbol(filename, linenr);

	return symbol_annotation(symbol);
}


void output_per_file_info(ostream & out, string const & filename,
			  u32 total_count_for_file)
{
	out << begin_comment << endl
	     << in_comment << "Total samples for file : "
	     << '"' << filename << '"'
	     << endl;
	out << in_comment << endl << in_comment
	    << counter_str(total_count_for_file, samples->samples_count())
	    << endl;
	out << end_comment << endl << endl;
}


string const line0_info(string const & filename)
{
	string annotation = source_line_annotation(filename, 0);
	if (trim(annotation, " \t:").empty())
		return string();

	string str = "<credited to line zero> ";
	str += annotation;
	return str;
}


void do_output_one_file(ostream & out, istream & in, string const & filename,
			bool header)
{
	u32 count = samples->samples_count(filename);

	if (header) {
		output_per_file_info(out, filename, count);
		out << line0_info(filename) << endl;
	}


	if (in) {
		string str;

		for (size_t linenr = 1 ; getline(in, str) ; ++linenr) {
			out << source_line_annotation(filename, linenr) << str
			    << source_symbol_annotation(filename, linenr)
			    << '\n';
		}

	} else {
		// FIXME : we have no input file : we just outputfooter
		// so on user can known total nr of samples for this source
		// later we must add code that iterate through symbol in this
		// file to output one annotation for each symbol. To do this we
		// need a select_symbol(filename); in profile_container which
		// fall back to the implementation in symbol_container
		// using a lazilly build symbol_map sorted by filename
		// (necessary functors already exist in symbol_functors.h)
	}

	if (!header) {
		output_per_file_info(out, filename, count);
		out << line0_info(filename) << endl;
	}
}


void output_one_file(istream & in, string const & filename,
		     bool output_separate_file)
{
	if (!output_separate_file) {
		do_output_one_file(cout, in, filename, true);
		return;
	}

	string out_filename = filename;

	size_t pos = out_filename.find(source_dir);
	if (pos == 0) {
		out_filename.erase(0, source_dir.length());
	} else if (pos == string::npos) {
		// filename is outside the source dir: ignore this file
		cerr << "opannotate: file "
		     << '"' << out_filename << '"' << " ignored" << endl;
		return;
	}

	out_filename = relative_to_absolute_path(output_dir + out_filename);

	if (create_path(out_filename.c_str())) {
		cerr << "unable to create directory: "
		     << '"' << dirname(out_filename) << '"' << endl;
		return;
	}

	// paranoid checking: out_filename and filename must be distinct file.
	if (is_files_identical(filename, out_filename)) {
		cerr << "input and output_filename are identical: "
		     << '"' << filename << '"'
		     << ','
		     << '"' << out_filename << '"'
		     << endl;
		return;
	}

	cout << "creating " << out_filename << endl;

	ofstream out(out_filename.c_str());
	if (!out) {
		cerr << "unable to open output file "
		     << '"' << out_filename << '"' << endl;
	} else {
		do_output_one_file(out, in, filename, false);
		output_info(out);
	}
}


void output_source(path_filter const & filter, bool output_separate_file)
{
	if (!output_separate_file)
		output_info(cout);

	vector<string> filenames =
		samples->select_filename(options::threshold);

	for (size_t i = 0 ; i < filenames.size() ; ++i) {
		if (!filter.match(filenames[i]))
			continue;

		ifstream in(filenames[i].c_str());

		if (!in) {
			// it is common to have empty filename due to the lack
			// of debug info (eg _init function) so warn only
			// if the filename is non empty. The case: no debug
			// info at all has already been checked.
			if (filenames[i].length())
				cerr << "opannotate (warning): unable to "
				     << "open for reading: "
				     << filenames[i] << endl;
		} 

		if (filenames[i].length()) {
			output_one_file(in, filenames[i], output_separate_file);
		}
	}
}


bool annotate_source(image_set const & images)
{
	annotation_fill = get_annotation_fill();

	bool output_separate_file = false;
	if (!source_dir.empty()) {
		output_separate_file = true;

		source_dir = relative_to_absolute_path(source_dir);
		if (source_dir.length() &&
		    source_dir[source_dir.length() - 1] != '/')
			source_dir += '/';
	}

	if (!output_dir.empty() || output_separate_file) {
		output_separate_file = true;

		output_dir = relative_to_absolute_path(output_dir);
		if (output_dir.length() &&
		    output_dir[output_dir.length() - 1] != '/')
			output_dir += '/';


		if (create_path(output_dir.c_str())) {
			cerr << "unable to create " << output_dir
			     << " directory: " << endl;
			return false;
		}
	}

	if (output_separate_file && output_dir == source_dir) {
		cerr << "You cannot specify the same directory for "
		     << "--output-dir and --source-dir" << endl;
		return false;
	}

	if (assembly) {
		image_set::const_iterator it;
		for (it = images.begin(); it != images.end(); ) {
			output_asm(it->first);

			it = images.upper_bound(it->first);
		}
	} else {
		output_source(file_filter, output_separate_file);
	}

	return true;
}


int opannotate(vector<string> const & non_options)
{
	handle_options(non_options);

	samples.reset(new profile_container(false, true, true));

	save_sample_file_header(*sample_file_partition);

	image_set images = populate_samples(*samples, *sample_file_partition,
					    false);
	annotate_source(images);

	return 0;
}

} // anonymous namespace


int main(int argc, char const * argv[])
{
	// set the invocation, for the file headers later
	for (int i = 0 ; i < argc ; ++i)
		cmdline += string(argv[i]) + " ";

	return run_pp_tool(argc, argv, opannotate);
}
